src/reader/mod.rs

```diff
@@ -124,33 +156,53 @@ impl TryInto<IndexReader> for IndexReaderBuilder {
 
 struct InnerIndexReader {
     num_searchers: usize,
-    searcher_pool: Pool<Searcher>,
     index: Index,
+    warming_state: WarmingState,
+    searcher_pool: Pool<Searcher>,
+    searcher_generation_counter: Arc<AtomicU64>,
+    searcher_generation_inventory: Inventory<SearcherIndexGeneration>,
 }
 
 impl InnerIndexReader {
     fn reload(&self) -> crate::Result<()> {
-        let segment_readers: Vec<SegmentReader> = {
-            let _meta_lock = self.index.directory().acquire_lock(&META_LOCK)?;
-            let searchable_segments = self.searchable_segments()?;
-            searchable_segments
-                .iter()
-                .map(SegmentReader::open)
-                .collect::<crate::Result<_>>()?
-        };
+        let (index_generation, segment_readers) = self.open()?;
         let schema = self.index.schema();
+
         let searchers: Vec<Searcher> = std::iter::repeat_with(|| {
-            Searcher::new(schema.clone(), self.index.clone(), segment_readers.clone())
+            Searcher::new(
+                schema.clone(),
+                self.index.clone(),
+                segment_readers.clone(),
+                index_generation.clone(),
+            )
         })
         .take(self.num_searchers)
         .collect::<io::Result<_>>()?;
+        self.warming_state
+            .warm_new_searcher_generation(&searchers[0])?;
         self.searcher_pool.publish_new_generation(searchers);
         Ok(())
     }
 
-    /// Returns the list of segments that are searchable
-    fn searchable_segments(&self) -> crate::Result<Vec<Segment>> {
-        self.index.searchable_segments()
+    fn open(&self) -> crate::Result<(TrackedObject<SearcherIndexGeneration>, Vec<SegmentReader>)> {
+        // Prevents segment files from getting deleted while we are in the process of opening them
+        let _meta_lock = self.index.directory().acquire_lock(&META_LOCK)?;
+        let searchable_segments = self.index.searchable_segments()?;
+        let generation = self
+            .searcher_generation_counter
+            .fetch_add(1, atomic::Ordering::Relaxed);
+        let searcher_generation = SearcherIndexGeneration::from_segment_metas(
+            searchable_segments.iter().map(|segment| segment.meta()),
+            generation,
+        );
```

Comment:
Minor, not a big deal either way: now that `SegmentReader` provides `delete_opstamp()`, we could  `from_segment_readers(&[SegmentReader])` using the `segment_readers` vec.

Reply:
Yes... By the way I had to add it, because it is necessary on the consumer side.

Reply:
Address.